File: leet/98m_ValidateBST.java
public class Solution {
    public boolean isValidBST(TreeNode node, long lo, long hi) {
        if(node==null) return true;
        if((long)node.val <= lo || (long)node.val >= hi)    return false;

        return  isValidBST(node.left, lo, (long)node.val) && isValidBST(node.right, (long)node.val, hi);
    }
    
    public boolean isValidBST(TreeNode root) {
        if(root==null)  return true;
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
}

File: leet/98m_ValidateBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def isValid(node, lo, hi):
            if not node:
                return True
            
            if node.val <= lo or node.val >= hi:
                return False
            
            return isValid(node.left, lo, node.val) and isValid(node.right, node.val, hi)
        
        return isValid(root, -pow(2, 32), pow(2, 32))
